File: library/src/main/java/cn/wandersnail/http/upload/SyncUploadWorker.java
package cn.wandersnail.http.upload;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import java.util.Map;

import cn.wandersnail.http.ConvertedResponse;
import cn.wandersnail.http.util.HttpUtils;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;

/**
 * 同步上传任务
 * <p>
 * date: 2019/8/23 20:38
 * author: zengfansheng
 */
public class SyncUploadWorker<T> {
    public ConvertedResponse<T> convertedResp;
    private final Handler handler = new Handler(Looper.getMainLooper());

    public SyncUploadWorker(UploadInfo<T> info, UploadProgressListener listener) {
        Retrofit.Builder builder = new Retrofit.Builder();
        if (info.client == null) {
            builder.client(HttpUtils.initHttpsClient(true, new OkHttpClient.Builder()).build());
        } else {
            builder.client(info.client);
        }
        UploadService service = builder.baseUrl(info.getBaseUrl())
                .build()
                .create(UploadService.class);
        MultipartBody.Builder bodyBuilder = new MultipartBody.Builder();
        bodyBuilder.setType(MultipartBody.FORM);
        if (info.paramParts != null) {
            for (Map.Entry<String, String> entry : info.paramParts.entrySet()) {
                bodyBuilder.addFormDataPart(entry.getKey(), entry.getValue());
            }
        }
        final InternalUploadListener localListener = new InternalUploadListener() {
            @Override
            public void onComplete(@NonNull FileInfo fileInfo) {
            }

            @Override
            public void onProgress(@NonNull FileInfo fileInfo, long progress, long max) {
                if (listener != null) {
                    handler.post(() -> listener.onProgress(fileInfo, progress, max));
                }
            }
        };
        for (FileInfo fileInfo : info.fileInfos) {
            ProgressRequestBody body = new ProgressRequestBody(fileInfo.getMediaType(), fileInfo, localListener);
            bodyBuilder.addFormDataPart(fileInfo.getFormDataName(), fileInfo.getFilename(), body);
        }
        Call<ResponseBody> call;
        if (info.headers == null || info.headers.isEmpty()) {
            call = service.upload(info.url, bodyBuilder.build());
        } else {
            call = service.upload(info.url, bodyBuilder.build(), info.headers);
        }
        convertedResp = new ConvertedResponse<>(call);
        try {
            Response<ResponseBody> response = call.execute();
            convertedResp.response = response;
            ResponseBody body = response.body();
            if (response.isSuccessful() && info.converter != null && body != null) {
                try {
                    convertedResp.convertedResponse = info.converter.convert(body);
                    body.close();
                } catch (Throwable t) {
                    convertedResp.convertError = t;
                }
            }
        } catch (Exception e) {
            //取消任务会抛异常
        }
    }
}

File: library/src/main/java/com/snail/network/upload/SyncUploadWorker.kt
package com.snail.network.upload

import com.snail.network.ConvertedResponse
import okhttp3.MediaType
import okhttp3.MultipartBody
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.net.URLEncoder
import java.util.concurrent.Executors

/**
 * 同步的上传执行者
 *
 * date: 2019/5/17 11:49
 * author: zengfansheng
 */
internal class SyncUploadWorker<T> @JvmOverloads constructor(info: UploadInfo<T>, listener: UploadProgressListener? = null) {
    val convertedResponse: ConvertedResponse<T>
    
    private val localListener = object : UploadProgressListener {
        private val executor = Executors.newSingleThreadExecutor()
        
        override fun onProgress(name: String, progress: Long, max: Long) {
            executor.execute { listener?.onProgress(name, progress, max) }
        }
    }
    
    init {
        val builder = Retrofit.Builder()
        if (info.client != null) {
            builder.client(info.client)
        }
        val service = builder.addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .baseUrl(info.baseUrl)
            .build()
            .create(UploadService::class.java)
        val bodyBuilder = MultipartBody.Builder()
        info.paramParts.entries.forEach {
            bodyBuilder.addFormDataPart(it.key, it.value)
        }
        info.fileParts.entries.forEach {
            val part = MultipartBody.Part.createFormData(it.key, URLEncoder.encode(it.value.name, "utf-8"),
                ProgressRequestBody(MediaType.parse("multipart/form-data"), it.key, it.value, localListener))
            bodyBuilder.addPart(part)
        }
        val call = service.uploadSync(info.url, bodyBuilder.build())
        convertedResponse = ConvertedResponse(call)
        try {
            val response = call.execute()
            convertedResponse.raw = response.raw()
            if (response.isSuccessful) {
                try {
                    convertedResponse.convertedBody = info.converter.convert(response.body())
                } catch (t: Throwable) {
                    convertedResponse.convertRrror = t
                }
            }
        } catch (e: Exception) {
            //取消任务会抛异常
        }
    }
}